### ISSUE TITLE
Remove type synonyms in instance head restriction

### ISSUE THREAD

**Issue by natefaubion** [purescript/purescript#3391]
A common pattern (for me at least)
```
type Environment = { ... }

instance askMyApp :: MonadAsk Environment MyApp where
  ...
```

But this doesn't work because type synonyms aren't allowed. Inlining the type is often not practical. The solutions are to newtype it, or unnecessarily use TypeEquals and coercions. I don't think this restriction accomplishes much. The only reason it exists is to temper expectations around allowable instances (type synonyms aren't nominal), but it does it poorly and the error does not even indicate why it's a problem.

**Comment by garyb** [comment]
I used to be against this, but I think I've come around to it.

**Comment by paf31** [comment]
The obvious place to allow it is under a fundep, right?

**Comment by natefaubion** [comment]
> The obvious place to allow it is under a fundep, right?

That's where it is most useful, I'd think. Doesn't that complicate synonym desugaring though? What exactly are we trying to guard against with this restriction and the error it presents? Are the cases already solved by other errors (orphan check)?

**Comment by hdgarrood** [comment]
It's worth noting that in GHC, `FlexibleInstances` implies `TypeSynonymInstances`: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-FlexibleInstances

We essentially have `FlexibleInstances` and `FlexibleContexts` in PureScript, too. In fact, we even have [`UndecidableInstances`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#undecidable-instances): we've dropped the requirement that instance resolution will provably terminate, and instead kept track of a certain amount of "fuel" to bail out after a given number of instance resolution steps.

So I agree with @natefaubion in that I think we should consider dropping the requirement totally: allow type synonyms to appear anywhere in both instances and contexts, desugar them before instance resolution, and use the existing instance rules to decide whether the resulting instances are valid.

**Comment by rhendric** [comment]
Fixed in #3539, I believe.
